blockdag/dagio.go

```diff
@@ -360,151 +270,135 @@ func (dag *BlockDAG) removeDAGState() error {
-				}
-			}
+		// Check to see if this node had been stored in the the block DB
+		// but not yet accepted. If so, add it to a slice to be processed later.
+		if node.status == statusDataStored {
+			unprocessedBlockNodes = append(unprocessedBlockNodes, node)
+			continue
+		}
 
-			// Add the node to its parents children, connect it,
-			// and add it to the block index.
-			node.updateParentsChildren()
+		// If the node is known to be invalid add it as-is to the block
+		// index and continue.
+		if node.status.KnownInvalid() {
 			dag.index.addNode(node)
+			continue
+		}
 
-			dag.blockCount++
+		if dag.blockCount == 0 {
+			if !node.hash.IsEqual(dag.dagParams.GenesisHash) {
+				return AssertError(fmt.Sprintf("initDAGState: Expected "+
+					"first entry in block index to be genesis block, "+
+					"found %s", node.hash))
+			}
+		} else {
+			if len(node.parents) == 0 {
+				return AssertError(fmt.Sprintf("initDAGState: Could "+
+					"not find any parent for block %s", node.hash))
+			}
 		}
 
-		// Load all of the known UTXO entries and construct the full
-		// UTXO set accordingly. Since the number of entries is already
-		// known, perform a single alloc for them versus a whole bunch
-		// of little ones to reduce pressure on the GC.
-		log.Infof("Loading UTXO set...")
+		// Add the node to its parents children, connect it,
+		// and add it to the block index.
+		node.updateParentsChildren()
+		dag.index.addNode(node)
 
-		utxoEntryBucket := dbTx.Metadata().Bucket(utxoSetBucketName)
+		dag.blockCount++
+	}
 
-		// Determine how many UTXO entries will be loaded into the index so we can
-		// allocate the right amount.
-		var utxoEntryCount int32
-		cursor := utxoEntryBucket.Cursor()
-		for ok := cursor.First(); ok; ok = cursor.Next() {
-			utxoEntryCount++
-		}
+	// Load all of the headers from the data for the known DAG
+	// and construct the block index accordingly. Since the
+	// number of nodes are already known, perform a single alloc
+	// for them versus a whole bunch of little ones to reduce
+	// pressure on the GC.
 
-		fullUTXOCollection := make(utxoCollection, utxoEntryCount)
-		for ok := cursor.First(); ok; ok = cursor.Next() {
-			// Deserialize the outpoint
-			outpoint, err := deserializeOutpoint(bytes.NewReader(cursor.Key()))
-			if err != nil {
-				return err
-			}
+	// Load all of the known UTXO entries and construct the full
+	// UTXO set accordingly. Since the number of entries is already
+	// known, perform a single alloc for them versus a whole bunch
+	// of little ones to reduce pressure on the GC.
+	log.Infof("Loading UTXO set...")
 
-			// Deserialize the utxo entry
-			entry, err := deserializeUTXOEntry(bytes.NewReader(cursor.Value()))
-			if err != nil {
-				return err
-			}
+	// Determine how many UTXO entries will be loaded into the index so we can
+	// allocate the right amount.
+	var utxoEntryCount int32
+	cursor, err := dbaccess.UTXOSetCursor(dbaccess.NoTx())
```

Comment:
I might be wrong, but it seems to me that not pre-allocating will have a lower overhead than double-iterating over the cursor, and this pre-allocation defenitely adds complexity.

Also, it's in the start-up, so as long as the overhead isn't overly terrible, I don't mind it too much.

On another note: maybe it's possible to ask levelDB to check for length, without actually iterating over the whole thing?

Reply:
> On another note: maybe it's possible to ask levelDB to check for length, without actually iterating over the whole thing?

Doesn't seem to support such a thing, I'm afraid.